host/http.go

```diff
@@ -244,33 +415,293 @@ func (h *jobAPI) RegisterRoutes(r *httprouter.Router) error {
+// avoid dropping logs or sending partial logs over two lines.
+//
+// An outline of the process:
+//
+// * parent receives a request to exec a new daemon
+//
+// * parent creates a control socket pair (via socketpair(2))
+//
+// * parent starts a child process, passing the API listener as FD 3, and a
+//   control socket as FD 4
+//
+// * parent closes its API listener FD, state DBs and log followers.
+//
+// * parent signals the child to resume by sending "resume" message to control
+//   socket, followed by any partial log buffers.
+//
+// * child receives resume request, opens state DBs, seeds the log followers
+//   with the partial buffers and starts serving API requests
+//
+// * child signals parent it is now serving requests by sending "ok" message to
+//   control socket
+//
+// * parent sends response to client and shuts down seconds later
+//
+func (h *Host) Update(cmd *host.Command) error {
+	log := h.log.New("fn", "Update")
+
+	// dup the listener so we can close the current listener but still be
+	// able continue serving requests if the child exits by using the dup'd
+	// listener.
+	log.Info("duplicating HTTP listener")
+	file, err := h.listener.(*net.TCPListener).File()
+	if err != nil {
+		log.Error("error duplicating HTTP listener", "err", err)
 		return err
 	}
-	shutdown.BeforeExit(func() { l.Close() })
+	defer file.Close()
 
-	r := httprouter.New()
+	// exec a child, passing the listener and control socket as extra files
+	log.Info("creating child process")
+	child, err := h.exec(cmd, file)
+	if err != nil {
+		log.Error("error creating child process", "err", err)
+		return err
+	}
+	defer child.CloseSock()
+
+	// close our listener and state DBs
+	log.Info("closing HTTP listener")
+	h.listener.Close()
+	log.Info("closing state databases")
+	if err := h.CloseDBs(); err != nil {
+		log.Error("error closing state databases", "err", err)
+		return err
+	}
 
-	r.POST("/attach", attach.ServeHTTP)
+	log.Info("closing logs")
+	buffers, err := h.CloseLogs()
+	if err != nil {
+		log.Error("error closing logs", "err", err)
+		return err
+	}
+
+	log.Info("resuming child process")
+	if resumeErr := child.Resume(buffers); resumeErr != nil {
+		log.Error("error resuming child process", "err", resumeErr)
+
+		// the child failed to resume, kill it and resume ourselves
+		log.Info("killing child process")
+		child.Kill()
+
+		log.Info("reopening logs")
+		if err := h.OpenLogs(buffers); err != nil {
+			log.Error("error reopening logs", "err", err)
+			return err
+		}
 
-	jobAPI := &jobAPI{
-		host:             h,
-		connectDiscoverd: connectDiscoverd,
+		log.Error("recreating HTTP listener")
+		l, err := net.FileListener(file)
+		if err != nil {
+			log.Error("error recreating HTTP listener", "err", err)
+			return err
+		}
+		h.listener = l
+
+		log.Info("reopening state databases")
+		if err := h.OpenDBs(); err != nil {
+			log.Error("reerror opening state databases", "err", err)
+			return err
```

Comment:
If this or any of the above fails, we should probably exit so that the process supervisor can attempt a restart, right?